### PR TITLE
Rework "Updates" table checkbox selection controls

### PR DESCRIPTION
**Pull request by janfaracik**
https://user-images.githubusercontent.com/43062514/177877649-53f09952-0736-4e9c-a089-2eb98e194281.mov

This PR adds a reusable table header checkbox component (_working title_ `"row-selection-controller"` - if anybody has a better name for it **please** share) which will select/unselect all checkboxes in a table. It's also extensible and can have custom selection options, e.g. "Compatible plugins" for the Updates page.

For the Updates page, this replaces the controls at the bottom of the page.

### Proposed changelog entries

* Rework "Updates" table checkbox selection controls.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")` if applicable.
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@jenkinsci/sig-ux 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6806"></a>

### CODE REVIEW

File: war/package.json
Comment (janfaracik) [inline war/package.json:32]: I've had to update `eslint` due to my use of optional chaining (e.g. `moreOptionsAllButton?.addEventListener`)

**Comment by timja** [comment]
Looks to be a z-index issue:
image

**Comment by timja** [comment]
Selecting compatible is a bit weird atm. It doesn't tick the top checkbox, when you click the top checkbox it does clear selection as expected but you don't get any button feedback

**Comment by timja** [comment]
Why does select none need to be a thing? shouldn't that be replaced by clicking the check box again, that's fairly standard behaviour

**Comment by janfaracik** [comment]
> Why does select none need to be a thing? shouldn't that be replaced by clicking the check box again, that's fairly standard behaviour
> 
> image

Clicking the checkbox does check/uncheck all. For the design I was essentially inspired by Gmail who do something similar - 
Screenshot 2022-07-08 at 11 28 41

**Comment by janfaracik** [comment]
> Selecting compatible is a bit weird atm. It doesn't tick the top checkbox, when you click the top checkbox it does clear selection as expected but you don't get any button feedback
> 
> From just clicking plugins one by one it does this: image
> 
> which I assume is what compatible should do

Hopefully I'm understanding you correctly - the header checkbox has three states (checked, indeterminate, none), if you click 'Compatible' and not all plugins are checked the header checkbox state will be indeterminate. If all plugins are checked when you click 'Compatible' then the header checkbox state will be checked.

Might have to look into reworking it if the behaviour isn't clear

**Comment by timja** [comment]
Yes but compatible didn't seem to be setting the half state of the top checkbox.

clicking them individually did set that correctly, selecting compatible didn't set any state on it which looked wrong

**Comment by yaroslavafenkin** [comment]
I observe some weird behaviour when there are no updates available.
I've checked the checkbox and now I cannot change the state:
image
Selecting any of the items in the dropdown or trying to uncheck does nothing.

**Review by timja — COMMENTED** [review]
all issues I found are fixed, let's run through ATH to be safe though

**Comment by NotMyFault** [comment]
> all issues I found are fixed, let's run through ATH to be safe though

ATH is green, one test failure remains on this PR at `hudson.PluginManagerInstalledGUITest.test_enable_disable_uninstall`.

**Review by timja — APPROVED** [review]
LGTM, I re-tested locally after all the merges

**Review by NotMyFault — APPROVED** [review]
Looks awesome, apologises for my conflicts caused 🥴

**Comment by uhafner** [comment]
🚀 Thanks a lot for fixing this usability problem! 🚀 

What are compatible plugins? Is this just a demo or what are the semantics of compatible?

**Comment by NotMyFault** [comment]
> 🚀 Thanks a lot for fixing this usability problem! 🚀
> 
> What are compatible plugins? Is this just a demo or what are the semantics of compatible?

That'd be the case if a plugin uses a higher jenkins.version than the one you have installed, has unmet or faulty third party dependencies afaik.

(Though I'm not too sure if that is still up with the times, the UC doesn't let you update plugins to a jenkins.version you don't use.)

**Comment by timja** [comment]
> That'd be the case if a plugin uses a higher jenkins.version than the one you have installed, has unmet or faulty third party dependencies afaik.
> 
> (Though I'm not too sure if that is still up with the times, the UC doesn't let you update plugins to a jenkins.version higher than your current version.)

No that's a different mechanism

This is https://wiki.jenkins-ci.org/display/JENKINS/Marking+a+new+plugin+version+as+incompatible+with+older+versions

`hpi.compatibleSinceVersion` is the plugin pom property.

I tested the functionality by installing the slack plugin version 1.0 and then checking for updates.

**Comment by uhafner** [comment]
I see. Since this option makes only sense if there are actually rows that are marked as incompatible I would suggest to not show this option if there are no such rows available.

**Comment by janfaracik** [comment]
I've updated the branch to hide the dropdown if all updates available are compatible (thanks @uhafner!) and I've also added a small tooltip the Compatible menu item,` "Selects plugin updates that are compatible with their current version"`. If anything/the label could be improved/changed let me know :)

Thanks

**Review by uhafner — APPROVED** [review]
I haven't looked at the code, but the UI looks good now!

**Comment by NotMyFault** [comment]
This PR is now ready for merge. We will merge it after ~24 hours if there is no negative feedback.
Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process.
Thanks!

**Comment by NotMyFault** [comment]
> > This PR is now ready for merge. We will merge it after ~24 hours if there is no negative feedback.
> > Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process.
> > Thanks!
> 
> Heya!
> 
> Is there a blocker on merging at the moment? Got a [few PRs](https://github.com/jenkinsci/jenkins/pulls?q=is%3Apr+is%3Aopen+label%3Aready-for-merge+author%3Ajanfaracik) that are ready to merge but haven't been yet.
> 
> Thanks :)

Not that I'm aware of. Speaking for the quoted comment, it's meant to start the timer, I'm not eligible to merge PRs once the grace period expired.

**Comment by janfaracik** [comment]
> > > This PR is now ready for merge. We will merge it after ~24 hours if there is no negative feedback.
> > > Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process.
> > > Thanks!
> > 
> > 
> > Heya!
> > Is there a blocker on merging at the moment? Got a [few PRs](https://github.com/jenkinsci/jenkins/pulls?q=is%3Apr+is%3Aopen+label%3Aready-for-merge+author%3Ajanfaracik) that are ready to merge but haven't been yet.
> > Thanks :)
> 
> Not that I'm aware of. Speaking for the quoted comment, it's meant to start the timer, I'm not eligible to merge PRs once the grace period expired.

Ah gotcha! Thanks @NotMyFault and @timja :)

**Comment by NotMyFault** [comment]
https://github.com/jenkinsci/jenkins/pull/6806/commits/86ad254472fc4cacfc58e5d6a5de2919a7422caf causes [JENKINS-69100](https://issues.jenkins.io/browse/JENKINS-69110) and undoes a lot of work done on this PR.

**Comment by Bananeweizen** [comment]
Just to want to mention that this change has one big drawback: You now have to scroll both to the top and the bottom of the page to actually do updates. Previously both UI elements for selection and triggering were next to each other. Maybe that can be improved by adding the checkbox a second time at the bottom of the table?

**Comment by uhafner** [comment]
Another option would be to use the table content as scrollable area and let the controls (and the Jenkins footer) always be visible at the screen bottom. That would make sense for a lot of other pages as well.

**Comment by daniel-beck** [comment]
@Bananeweizen Please file this as an issue in Jira, or file a pull request implementing your suggestion.

**Pull request merged**
